src/spotify_to_tidal/sync.py

```diff
@@ -172,133 +181,138 @@ def get_tracks_from_spotify_playlist(spotify_session: spotipy.Spotify, spotify_p
     # get all the remaining tracks in parallel
     if results['next']:
         offsets = [ results['limit'] * n for n in range(1, math.ceil(results['total']/results['limit'])) ]
-        extra_results = call_async_with_progress(_get_tracks_from_spotify_playlist, offsets, "",
-                                                 min(len(offsets), 10), spotify_session=spotify_session, playlist_id=spotify_playlist["id"])
+        extra_results = await atqdm.gather( *[asyncio.to_thread(_get_tracks_from_spotify_playlist, offset, spotify_session=spotify_session, playlist_id=spotify_playlist["id"]) for offset in offsets ] )
         for extra_result in extra_results:
             output.extend([r['track'] for r in extra_result['items'] if r['track'] is not None])
     return output
 
-class TidalPlaylistCache:
-    def __init__(self, playlist: tidalapi.Playlist):
-        self._data = playlist.tracks()
-
-    def _search(self, spotify_track: t_spotify.SpotifyTrack):
-        ''' check if the given spotify track was already in the tidal playlist.'''
-        results = []
-        for tidal_track in self._data:
+def populate_track_match_cache(spotify_tracks: List[t_spotify.SpotifyTrack], tidal_tracks: List[tidalapi.Track]):
+    """ Populate the track match cache with all the existing tracks in Tidal playlist corresponding to Spotify playlist """
+    def _populate_one_track(spotify_track: t_spotify.SpotifyTrack) -> bool:
+        for idx, tidal_track in list(enumerate(tidal_tracks)):
             if match(tidal_track, spotify_track):
-                return tidal_track
-        return None
-
-    def search(self, spotify_session: spotipy.Spotify, spotify_playlist):
-        ''' Add the cached tidal track where applicable to a list of spotify tracks '''
-        results = []
-        cache_hits = 0
-        work_to_do = False
-        spotify_tracks = get_tracks_from_spotify_playlist(spotify_session, spotify_playlist)
-        for track in spotify_tracks:
-            cached_track = self._search(track)
-            if cached_track:
-                results.append( (track, cached_track) )
-                cache_hits += 1
-            else:
-                results.append( (track, None) )
-        return (results, cache_hits)
+                track_match_cache.insert((spotify_track['id'], tidal_track.id))
+                return True
```

Comment:
I removed the call to `pop()` here as I observed an edge case where sometimes you have multiple Spotify tracks in a playlist that have different IDs but reference the same song. For example a single vs. album version. In this case, both of these Spotify tracks would map to the same Tidal track, and so if you pop the Tidal track on first match, you end up searching for these outlier tracks on every sync even if there are no new tracks.